Update management change to support Dynamic group, Prepost and Reboot setting

jemex opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description

Azure automation update management change to support the following new features
       Dynamic grouping
       Pre-Post script
      Reboot Setting
cmdlet Design review : https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/230

## Checklist

- [x ] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md)
- [x ] The title of the PR is clear and informative
- [x ] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#updating-the-change-log)
- [x ] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (Microsoft internal partners only)
    - [ ] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/help-generation.md)

cormacpayne commented:
@jemex CI failed for the following reasons:

(1) StaticAnalysis -- the parameter `-AzureQueries` breaks the ["singular noun" rule](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/azure-powershell-design-guidelines.md#singularity) in PowerShell. To fix this, the parameter should be renamed to `-AzureQuery`.

(2) Test failures -- the following test failed during CI:

- TestImportRunbookPowerShell
- TestImportAndDeleteRunbookGraphical
- TestCreateRunbookGraph

Please make sure that these tests pass locally in playback, and re-record them if you continue to see failures.

jemex commented:
Hi @cormacpayne , thank you so much for the pointer you gave me the Ci build issue . I have fixed it and it has passed now.

jemex commented:
How to tell the failure reason for the CI? what is the issue?

cormacpayne commented on src/Automation/Automation/Cmdlet/UpdateManagement/NewAzureAutomationSoftwareUpdateConfiguration.cs:
```diff
@@ -40,15 +42,39 @@ public class NewAzureAutomationSoftwareUpdateConfiguration : AzureAutomationBase
         [Parameter(ParameterSetName = AutomationCmdletParameterSets.Linux, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "Indicates that the software update configuration targeting Linux operating system machines.")]
         public SwitchParameter Linux { get; set; }
 
+        [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "Indicates that the software update configuration will Only Reboot the machines.")]
```
@jemex you can remove `ValueFromPipelineByPropertyName` from all `SwitchParameter` parameters

cormacpayne commented on src/Automation/Automation/Cmdlet/UpdateManagement/NewAzureAutomationSoftwareUpdateConfiguration.cs:
```diff
@@ -40,15 +42,39 @@ public class NewAzureAutomationSoftwareUpdateConfiguration : AzureAutomationBase
+        public string PostTaskRunbookName { get; set; }
+
+        [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "Pre task parameter.")]
+        public Hashtable PreTaskRunbookParameters { get; set; }
```
@jemex another issue you'll see in the `SignatureIssues.csv` file is about [noun singularity in parameter names](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/azure-powershell-design-guidelines.md#singularity) -- from that rule:

_Avoid using plural names for parameters whose value is a single element. This includes parameters that take arrays or lists because the user might supply an array or list with only one element._

_Plural parameter names should be used only in those cases where the value of the parameter is always a multiple-element value. In these cases, the cmdlet should verify that multiple elements are supplied, and the cmdlet should display a warning to the user if multiple elements are not supplied._

jemex replied:
@cormacpayne Thank you so much for the explanation. It make sense. I changed it to singular

cormacpayne commented on src/Automation/Automation/Model/UpdateManagement/TagOperators.cs:
```diff
@@ -0,0 +1,14 @@
+﻿using System;
```
@jemex please add a license header to this file

cormacpayne commented on src/Automation/Automation/Properties/Resources.resx:
```diff
@@ -490,4 +490,7 @@
+  <data name="SoftwareUpdateConfigurationInvalidTagFormat" xml:space="preserve">
+    <value>The Tag format is invalid. Valid Tag format is hashtable with string key and value of String or array of string.</value>
+  </data>
 </root>
```
@jemex (I can't comment on the file itself) why was the below `PS-VSPrompt.lnk` file changed?

jemex replied:
@cormacpayne  I didn't make any intentional change to this file. I will revert the file

jemex commented on src/Automation/Automation.Test/ScenarioTests/UpdateDynamicGroupPrePostTests.ps1:
```diff
@@ -0,0 +1,221 @@
+                                                                -Name $Name
+        $state = $suc.ProvisioningState
+        Write-Output "SoftwareUpdateConfiguration Provisioning state: $state"
+        sleep -Seconds 5
```
this whole function is not needed. removed

jemex commented on src/Automation/Automation.Test/ScenarioTests/ModuleTests.cs:
```diff
@@ -53,7 +53,11 @@ public void NewModule()
+        /* Message: System.Exception : Subscription Id is not present in the 
+         * recorded mock or in connection string (e.g. SubscriptionId=<subscriptionId>).
+         */
+        [Fact(Skip = "Test does not work on playback mode ")]
```
the test works on recording mode. How ever it failed in replaying mode with that error.  I was not sure why the CI build was failing and thought that could be the reason.  now reverted back this change.

jemex commented on src/Automation/Automation/Cmdlet/UpdateManagement/NewAutomationUpdateManagementAzureQuery .cs:
```diff
@@ -0,0 +1,99 @@
+    using TagHelper = Microsoft.Azure.Commands.ResourceManager.Common.Tags;
+    using System.Collections.Generic;
+
+    [Cmdlet(VerbsCommon.New, ResourceManager.Common.AzureRMConstants.AzurePrefix + "AutomationUpdateManagementAzureQuery")]
```
@cormacpayne Thank you so much for explaining what the issue is and how to fix it. I added suppression.

jemex commented on src/Automation/Automation.Test/ScenarioTests/UpdateDynamicGroupPrePostTests.ps1:
```diff
@@ -0,0 +1,221 @@
+$rg = "mms-wcus"
+$aa = "JemalOMSAutomation"
+$azureVMIdsW = @(
+        "/subscriptions/cd45f23b-b832-4fa4-a434-1bf7e6f14a5a/resourceGroups/mms-wcus/providers/Microsoft.Compute/virtualMachines/JemalCmdlet1",
```
@cormacpayne    the test needs prepacked Vms that has all the requirements  for the feature to run successfully. The hardcoded in the test is the azure resource id for these vms. It will be little difficult to prepare these Vms with all the requirements in the test it self.

if the user doesn't have permission on the resources and would like to re record the test, the user can change the resources to the one he has permission and record the test.   It is not that difficult to change the resource ids  and record the test

cormacpayne replied:
@jemex for tests that require additional setup, if there's no way around hardcoding resource ids, then you should leave a comment above this section outlining what a user will need to do (either in the portal or using commands) in order to get this test working correctly

jemex commented:
@cormacpayne , I addressed all the feedback that could resolve the CI build issue. However, CI still seems to fail. I couldn't understand the build issue. Can you help me to direct the main cause of the issue, please?,

cormacpayne commented on src/Automation/Automation.Test/ScenarioTests/UpdateDynamicGroupPrePostTests.cs:
```diff
@@ -0,0 +1,72 @@
+    using Microsoft.WindowsAzure.Commands.ScenarioTest;
+    using Xunit;
+
+    public class UpdateDynamicGroupPrePostTests : AutomationScenarioTestsBase
```
@jemex the build is failing partially because you are inheriting from this `AutomationScenarioTestBase` class that doesn't exist anymore -- you should be inheriting from `AutomationTestRunner`

cormacpayne commented on src/Automation/Automation/help/New-AzAutomationUpdateManagementAzureQuery.md:
```diff
@@ -0,0 +1,161 @@
+# New-AzAutomationUpdateManagementAzureQuery
+
+## SYNOPSIS
+{{Fill in the Synopsis}}
```
@jemex please fill out all template sections provided in this file

jemex replied:
@cormacpayne The template is now filled

cormacpayne commented on src/Automation/Automation.Test/ScenarioTests/UpdateDynamicGroupPrePostTests.cs:
```diff
@@ -0,0 +1,72 @@
+    using Microsoft.WindowsAzure.Commands.ScenarioTest;
+    using Xunit;
+
+    public class UpdateDynamicGroupPrePostTests : AutomationScenarioTestsBase
```
@jemex the build is failing partially because you are inheriting from this `AutomationScenarioTestBase` class that doesn't exist anymore -- you should be inheriting from `AutomationTestRunner`

jemex replied:
@cormacpayne , Thank you so much, I wasn't able o figure it out. now is inheriting correctly

cormacpayne commented on src/Automation/Automation/help/New-AzAutomationUpdateManagementAzureQuery.md:
```diff
@@ -0,0 +1,161 @@
+---
+external help file: Microsoft.Azure.PowerShell.Cmdlets.Automation.dll-Help.xml
+Module Name: Az.Automation
+online version:
```
@jemex you will need to add the correct URL for this cmdlet as it would appear on docs. Here's the message from the `ValidateHelpIssues.csv` file in the build artifacts:

```
Online version in the header of the file is incorrect. The corresponding URL should be the following: https://docs.microsoft.com/en-us/powershell/module/az.automation/new-azautomationupdatemanagementazurequery
```

jemex replied:
@cormacpayne Thank you for the tip, correct url is added

vladimir-shcherbakov commented:
@jemex Please fix the ci build. This is the issue: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell/11209/artifact/artifacts/ValidateHelpIssues.csv/*view*/

vladimir-shcherbakov commented:
@jemex  Please fix the Static Analysis issues: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell/11224/artifact/artifacts/BreakingChangeIssues.csv/*view*/

cormacpayne commented on src/Accounts/Accounts/ChangeLog.md:
```diff
@@ -29,6 +29,10 @@
 
 ## Version 1.3.0
 * Update to latest version of ClientRuntime
+* Azure automation update management change to support the following new features :
```
@jemex the text you are adding to this file should go under the `## Upcoming Release` header -- looks like this moved down when the last release occurred

cormacpayne commented on src/Automation/Automation/Cmdlet/UpdateManagement/NewAzureAutomationSoftwareUpdateConfiguration.cs:
```diff
@@ -24,47 +24,73 @@ namespace Microsoft.Azure.Commands.Automation.Cmdlet.UpdateManagement
     [OutputType(typeof(SoftwareUpdateConfiguration))]
     public class NewAzureAutomationSoftwareUpdateConfiguration : AzureAutomationBaseCmdlet
     {
-        [Parameter(Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "Schedule object used for software update configuration.")]
```
@jemex is there any reason that the existing `ValueFromPipelineByPropertyName` property is being removed from these parameters?

jemex replied:
@cormacpayne I removed it because in one of your feedback, you mentioned that it is ok to remove it. It looks like I misunderstood the feedback you gave earlier. I will add it back. is that why the CI build failing?
